### ISSUE TITLE
Diff Mode often has UndefinedClass errors

### ISSUE THREAD

**Issue by beberlei** [vimeo/psalm#4402]
Trying out psalm 4 with the new diff mode, but it often throws `UndefinedClass` errors, ususally classes that are tangitally related to the file that was changed. For example a class is changed and Psalm cannot find the interface it implements, but also seemingly unrelated classes. This is on a huge project, so I am not sure what debug infos to provide to get ahold of the problem.

**Comment by muglug** [comment]
> ususally classes that are tangitally related to the file that was changed

Can the autoloader figure out where the classes are?

**Comment by beberlei** [comment]
@muglug yes I believe so, when i run without --no-diff it can find them no problem. Depening on changes in the files, other classes cannot be found. Autoloading related, we do use the Composer monorepo plugin, maybe that is the culprit? https://github.com/beberlei/composer-monorepo-plugin

**Comment by muglug** [comment]
That absolutely sounds like the culprit.

If you're using Psalm on the all the files in a given repo, they don't _have_ to be autoloadable (because it will load all the classes it can find in those directories).

But if you're using Psalm in diff mode, those files _do_ have to be autoloadable.

By default Psalm uses composer's autoloader, but you can specify one yourself [with this config option](https://psalm.dev/docs/running_psalm/configuration/#autoloader).

**Comment by muglug** [comment]
Psalm _could_ allow this by constructing its own map of classes to paths when it first scans – like a last-known-location sort of thing.

**Comment by beberlei** [comment]
@muglug thank you for the tip about the own autoloader, this will allow me to work around this. The monorepo plugis obviously the reason for this problem.
